CMakeLists.txt

```diff
@@ -60,9 +60,18 @@ endif()
 
 if(VOLK_PULL_IN_VULKAN)
   # If CMake has the FindVulkan module and it works, use it.
-  # Otherwise try the environment variable.
   find_package(Vulkan QUIET)
-  if(TARGET Vulkan::Vulkan)
+
+  # Try an explicit CMake variable first, then any Vulkan targets
+  # discovered by FindVulkan.cmake, then the $VULKAN_SDK environment
+  # variable if nothing else works.
+  if(VULKAN_HEADERS_INSTALL_DIR)
```

Comment:
How does this work - what is setting this variable?

Reply:
(I'd expect this to be defined in CMakeLists.txt as an option but that doesn't seem to be happening here)

Reply:
Perhaps it should be?... Looking at the other **Vulkan SDK** repositories, they uniformly consume intra-SDK dependencies by receiving an otherwise mostly undocumented pointer to the dependency's `install` directory.  But the CMake code involved is pretty ancient and may not reflect current best practices.

During an SDK build, the `cmake` executable receives `-DVAR=VALUE` for each declared intra-SDK dependency, on an invoked command line, during generation.

We're happy to use any other such mechanism you prefer; we cannot use the current `$VULKAN_SDK` environment variable because the existing code will check for the presence of Vulkan installed via former SDK first, and neither want to accidentally build from an SDK's headers, nor to change current behavior by re-ordering the header search.  A new variable seemed most likely not to break other builds.

Reply:
I'd need to see what other projects do. If VULKAN_SDK environment variable took precedence over the Vulkan CMake module, would this work for you as well?

Reply:
Looking into this further, I guess CMake will happily override any internal variable, they just wouldn't show up in the configuration help. So this works, but maybe isn't best practice. Since there's already two official configuration methods though I can merge this as is for now and just have it be the internal hook.

That said, I would prefer simply reusing the existing VULKAN_SDK env if checking it before checking the CMake module works fine, there's no indication that the current order is better. If this doesn't work then I will merge the extra setting.

Reply:
Thanks, that would be great. The issue is we build this for the SDK... and we don't want to use the last SDK's headers, so @lunarpapillo is looking for a clean way to set that without having to clear the VULKAN_SDK environment variable if an SDK is already installed on the build machine.

Reply:
Ok, so just to make sure I understand it, you need to override not just the Vulkan CMake library discovery, but *also* bypass the use of VULKAN_SDK environment variable because it might not be pointing to the SDK that you want to test with / build against either?

Reply:
Correct me if I'm wrong: if there's a Vulkan SDK installed at all, `find_package(Vulkan QUIET)` will resolve to that, and `if(TARGET Vulkan::Vulkan)` will resolve to `True`, and that SDK will be used for compiling **Volk**, and the `$ENV{VULKAN_SDK}` branch will never be hit.  So as far as **Volk** is concerned, whatever value is in `$VULKAN_SDK` is ignored in this case.

Now, at least some versions of `FindVulkan` do in fact refer to `$ENV{VULKAN_SDK}` (as of **CMake 3.18**, if I'm not mistaken), but older versions of **CMake** could quietly find a different set of headers than was intended, correct?  In this case the builder would believe they were building with the correct version of headers, but the build would be incorrect in such a way that would be difficult to detect.

I just need to reliably define the headers to be used, without being sensitive to whatever is installed on the system (whether an old version of **CMake** or an old version of the **Vulkan SDK**).

We currently unofficially support SDK building with **CMake 3.10** and later (to support older distributions).  We could propose raising the minimum to **CMake 3.18** so we can include **Volk** reliably, but historically such discussions have met with vocal resistance.

Does that make sense?  Am I missing anything critical?

Reply:
> I would prefer simply reusing the existing VULKAN_SDK env if checking it before checking the CMake module works fine, there's no indication that the current order is better.

I think that can work reliably enough; but it is provably a change in behavior in some cases (i.e. if you have a **Vulkan SDK** installed and a version of **CMake** before 3.18), which could feasibly break downstream users.

If you're comfortable with that possibility, we can make that work.

Reply:
Let's just use this PR for now; we can refine all of this a little bit in the future, eg I don't think it's actually a meaningful compat problem for existing users if VULKAN_SDK env var takes priority, but at the same time it's more "clean" to prefer the CMake package route (which I think is why the order of checks is what it is). For now a separate option seems reasonable.